packages/langium/src/grammar/langium-grammar-validator.ts

```diff
@@ -438,6 +443,27 @@ export class LangiumGrammarValidator {
             accept('error', 'Parser rules cannot be used for cross references.', { node: reference.terminal, property: 'rule' });
         }
     }
+
+    checkCrossRefType(reference: ast.CrossReference, accept: ValidationAcceptor): void {
+        const issue = this.checkReferenceToRuleButNotType(reference.type);
+        if (issue) {
+            accept('error', issue, { node: reference, property: 'type' });
+        }
+    }
+
+    checkAtomTypeRefType(atomType: ast.AtomType, accept: ValidationAcceptor): void {
+        const issue = this.checkReferenceToRuleButNotType(atomType.refType);
+        if (issue) {
+            accept('error', issue, { node: atomType, property: 'refType' });
+        }
+    }
+
+    protected checkReferenceToRuleButNotType(type: Reference<ast.AbstractType>): string | undefined {
+        if(type && ast.isParserRule(type.ref) && !isDataTypeRule(type.ref) && type.ref.type) {
+            return `Use the rule type '${type.ref.type.name}' instead of the typed rule '${type.ref.name}' for cross references.`;
```

Comment:
The error message looks to me a bit confusing. I'd propose such one:
`The rule '${type.ref.name}' has a type that doesn't match with its name. Use the type '${type.ref.type.name}' when want to reference on a rule '${type.ref.name}'.`

Reply:
> The rule '${type.ref.name}' has a type that doesn't match with its name.

Is that the real problem why we do the validation? I thought it is more like we need to reference the Type if it is explicitly defined. And we actually do link against the type even if it has the same name.

```js
Definition returns Definition:
	name=ID;
```

Reply:
@Pluralia 
I will merge the current state if you don't mind. And I also think that someone should look over the messages we are currently producing and check for plausibility, comprehensibility and spelling :)

Reply:
> Is that the real problem why we do the validation?

Basically, yes.
The real problem is cross-references can be only on types, but not on rules, so, when we want to make a cross-reference on a rule, we have to know a type of the rule. If a rule has an explicitly defined type -- the rule type is obvious. If no explicitly defined type -- the rule type is something, what is inferred from the rule and has the same name that the rule has. So, by the phrase `The rule '${type.ref.name}' has a type that doesn't match with its name.` I wanted to say, that the rule has an explicitly defined type that is not inferred from the rule.

Your example with `Definition` is nice and shows that this phrase is not completely correct. A rule type can really have the same name as the rule when an explicitly defined type has the same name as the rule. However, when we reference on such rule like `Definition`, we use the `Definition` word in cross-reference, and it anyway will be recognized as a reference on the return type, so, the error will not rise. The error will rise exactly in cases when the phrase is correct.